exec/java-exec/src/main/java/org/apache/drill/exec/expr/fn/DrillFunctionRegistry.java

```diff
@@ -218,4 +302,141 @@ private void registerOperatorsWithoutInference(DrillOperatorTable operatorTable)
       }
     }
   }
+
+  /**
+   * Function registry holder. Stores function implementations by jar name, function name.
+   */
+  private class GenericRegistryHolder<T, U> {
+    private final ReadWriteLock readWriteLock = new ReentrantReadWriteLock();
+    private final AutoCloseableLock readLock = new AutoCloseableLock(readWriteLock.readLock());
+    private final AutoCloseableLock writeLock = new AutoCloseableLock(readWriteLock.writeLock());
+
+    // jar name, Map<function name, List<signature>
+    private final Map<T, Map<T, List<T>>> jars;
+
+    // function name, Map<signature, function holder>
+    private final Map<T, Map<T, U>> functions;
+
+    public GenericRegistryHolder() {
+      this.functions = Maps.newHashMap();
+      this.jars = Maps.newHashMap();
+    }
+
+    public void addJar(T jName, Map<T, Pair<T, U>> sNameMap) {
+      try (AutoCloseableLock lock = writeLock.open()) {
+        Map<T, List<T>> map = jars.get(jName);
+        if (map != null) {
+          removeAllByJar(jName);
+        }
```

Comment:
Redundant. removeAllByJar does the check to see if the jar exists. Just do:

removeAllByJar(jName);

WIthout the duplicate lookup.